syntax = "proto3";

package brain.state.v1;

import "brain/shared/v1/envelope.proto";
import "google/protobuf/timestamp.proto";

// Authority service for vault notes and directory operations.
//
// Each RPC uses operation-specific envelope-like request/response messages.
service VaultAuthorityService {
  rpc ListDirectory(ListDirectoryRequest) returns (ListDirectoryResponse);
  rpc CreateDirectory(CreateDirectoryRequest) returns (CreateDirectoryResponse);
  rpc DeleteDirectory(DeleteDirectoryRequest) returns (DeleteDirectoryResponse);
  rpc PutNote(PutNoteRequest) returns (PutNoteResponse);
  rpc GetNote(GetNoteRequest) returns (GetNoteResponse);
  rpc DeleteNote(DeleteNoteRequest) returns (DeleteNoteResponse);
  rpc AppendNote(AppendNoteRequest) returns (AppendNoteResponse);
  rpc SearchNotes(SearchNotesRequest) returns (SearchNotesResponse);
}

// Canonical note path reference in the vault namespace.
message NoteRef {
  string note_path = 1;
}

// Canonical directory path reference in the vault namespace.
message DirectoryRef {
  string directory_path = 1;
}

// Materialized note metadata and content.
message NoteRecord {
  NoteRef ref                              = 1;
  string content                           = 2;
  google.protobuf.Timestamp created_at     = 3;
  google.protobuf.Timestamp updated_at     = 4;
}

// Search match payload for note search operations.
message NoteSearchMatch {
  NoteRef ref          = 1;
  repeated string hits = 2;
}

message ListDirectoryRequest {
  brain.shared.v1.EnvelopeMeta meta = 1;
  DirectoryRef ref                    = 2;
}

message ListDirectoryResponse {
  brain.shared.v1.EnvelopeMeta meta           = 1;
  repeated DirectoryRef directory_paths         = 2;
  repeated brain.shared.v1.ErrorDetail errors = 3;
}

message CreateDirectoryRequest {
  brain.shared.v1.EnvelopeMeta meta = 1;
  DirectoryRef ref                    = 2;
  bool recursive                      = 3;
}

message CreateDirectoryResponse {
  brain.shared.v1.EnvelopeMeta meta = 1;
  DirectoryRef ref                    = 2;
  repeated brain.shared.v1.ErrorDetail errors = 3;
}

message DeleteDirectoryRequest {
  brain.shared.v1.EnvelopeMeta meta = 1;
  DirectoryRef ref                    = 2;
  bool recursive                      = 3;
  bool missing_ok                     = 4;
}

message DeleteDirectoryResponse {
  brain.shared.v1.EnvelopeMeta meta = 1;
  bool deleted                        = 2;
  repeated brain.shared.v1.ErrorDetail errors = 3;
}

message PutNoteRequest {
  brain.shared.v1.EnvelopeMeta meta = 1;
  NoteRef ref                         = 2;
  string content                      = 3;
}

message PutNoteResponse {
  brain.shared.v1.EnvelopeMeta meta = 1;
  NoteRecord note                     = 2;
  repeated brain.shared.v1.ErrorDetail errors = 3;
}

message GetNoteRequest {
  brain.shared.v1.EnvelopeMeta meta = 1;
  NoteRef ref                         = 2;
}

message GetNoteResponse {
  brain.shared.v1.EnvelopeMeta meta = 1;
  NoteRecord note                     = 2;
  repeated brain.shared.v1.ErrorDetail errors = 3;
}

message DeleteNoteRequest {
  brain.shared.v1.EnvelopeMeta meta = 1;
  NoteRef ref                         = 2;
  bool missing_ok                     = 3;
}

message DeleteNoteResponse {
  brain.shared.v1.EnvelopeMeta meta = 1;
  bool deleted                        = 2;
  repeated brain.shared.v1.ErrorDetail errors = 3;
}

message AppendNoteRequest {
  brain.shared.v1.EnvelopeMeta meta = 1;
  NoteRef ref                         = 2;
  string content                      = 3;
}

message AppendNoteResponse {
  brain.shared.v1.EnvelopeMeta meta = 1;
  NoteRecord note                     = 2;
  repeated brain.shared.v1.ErrorDetail errors = 3;
}

message SearchNotesRequest {
  brain.shared.v1.EnvelopeMeta meta = 1;
  string query                        = 2;
  DirectoryRef scope                  = 3;
  int32 limit                         = 4;
}

message SearchNotesResponse {
  brain.shared.v1.EnvelopeMeta meta = 1;
  repeated NoteSearchMatch matches    = 2;
  repeated brain.shared.v1.ErrorDetail errors = 3;
}
